### PR TITLE
add mkldnn batch_norm backward

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37218 BFloat16: enable prepacked weights's inference
* #37215 BFloat16: add explicit dtype support for to_mkldnn and to_dense
* #37199 add mkldnn relu backward and reshape backward
* **#37147 add mkldnn batch_norm backward**
* #37146 add mkldnn pooling backward
* #36122 add mkldnn linear backward
* #36121 enable mkldnn conv2d backward to support mkldnn tensor input

Differential Revision: [D22440965](https://our.internmc.facebook.com/intern/diff/D22440965)

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 1]
```diff
@@ -39,35 +53,86 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm(
...
+  ideep::tensor& x = itensor_from_mkldnn(input);
+  ideep::tensor w = itensor_from_tensor(weight);
+  ideep::tensor& m = itensor_from_mkldnn(save_mean);
+  ideep::tensor& v = itensor_from_mkldnn(save_invstd);
+
+  ideep::tensor gradx, gradw, gradb;
+  ideep::batch_normalization_backward::compute(
+      x, m, v, grady, w, gradx, gradw, gradb, eps);
+
+  if (weight.is_mkldnn()) {
```
Comment (albanD) [inline aten/src/ATen/native/mkldnn/Normalization.cpp:126]: There seems to be a lot of assumptions on the input types here. Can we have the corresponding checks both for the forward and backward functions?

File: test/test_mkldnn.py [hunk 2]
```diff
@@ -260,20 +260,40 @@ def test_adaptive_avg_pool2d(self):
...
+        for train in [True, False]:
+            # TODO: support none affine
+            for affine in [True]:
+                for track_running_stats in [True, False]:
+                    bn = torch.nn.BatchNorm2d(
+                        3,
+                        affine=affine,
+                        track_running_stats=track_running_stats).float().train(train)
+                    if (train or not track_running_stats):
+                        mkldnn_bn = copy.deepcopy(bn)
```
Comment (albanD) [inline test/test_mkldnn.py:274]: Could you explain why in this case you don't send the module to mkldnn?

File: test/test_mkldnn.py [hunk 3]
```diff
@@ -260,20 +260,40 @@ def test_adaptive_avg_pool2d(self):
...
+                        x2.requires_grad_()
+                    y1 = bn(x1)
+                    y2 = mkldnn_bn(x2).to_dense()
+                    if train:
+                        loss1 = y1.sum()
+                        loss2 = y2.sum()
+                        loss1.backward()
+                        loss2.backward()
+                        self.assertEqual(x1.grad, x2.grad.to_dense())
+                        np.testing.assert_allclose(
```
Comment (albanD) [inline test/test_mkldnn.py:290]: nit: you can use `self.assertEqual(t1, t2, rtol=xx, atol=xx)`.

File: test/test_mkldnn.py [hunk 4]
```diff
@@ -260,20 +260,40 @@ def test_adaptive_avg_pool2d(self):
...
+                            bn.weight.grad, mkldnn_bn.weight.grad, rtol=1e-3, atol=1e-3)
```
Comment (albanD) [inline test/test_mkldnn.py:291]: Why is the tolerance so high here give that you assert that y1, y2 are exactly equal below?

File: test/test_mkldnn.py [hunk 5]
```diff
@@ -260,20 +260,40 @@ def test_adaptive_avg_pool2d(self):
...
 
-            self._test_serialization(mkldnn_bn, (x.to_mkldnn(),))
-            self._test_tracing(mkldnn_bn, (x.to_mkldnn(),))
+                    self.assertEqual(y1, y2)
```
Comment (albanD) [inline test/test_mkldnn.py:293]: Could you also check that the running stats have been updated properly?

**Review by albanD — CHANGES_REQUESTED** [review]
Thanks for the PR!
Just some comments on the test and missing checks.

File: aten/src/ATen/native/mkldnn/Normalization.cpp (hunk 1 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Normalization.cpp:126]: Now, just using second path.

File: test/test_mkldnn.py (hunk 2 above)
Comment (XiaobingSuper) [reply test/test_mkldnn.py:274]: For training case, module's parameters awalys a dense tensor, not need to call **mkldnn_utils.to_mkldnn**

File: test/test_mkldnn.py (hunk 5 above)
Comment (XiaobingSuper) [reply test/test_mkldnn.py:292]: done.

File: test/test_mkldnn.py (hunk 4 above)
Comment (XiaobingSuper) [reply test/test_mkldnn.py]: diff_weight is computed by `Sum_over_MB*H*W(diff_dst[i] * x_normalized[i])`, the summation order of float number may get different number, MKLDNN will the whole job to a piece job for using muti-thread, the order sum may different with the native path, I also test cuda case, there also has same problem. but for y1, y2, there just has element wise operation, there has't big difference for MKLDNN and native path.

File: test/test_mkldnn.py (hunk 3 above)
Comment (XiaobingSuper) [reply test/test_mkldnn.py]: done.

File: test/test_mkldnn.py (hunk 4 above)
Comment (albanD) [reply test/test_mkldnn.py]: But given that you work with small Tensors and the only problem is ordering, a tolerance of 1e-5 should be enough no?

File: test/test_mkldnn.py (hunk 2 above)
Comment (albanD) [reply test/test_mkldnn.py:274]: Ho ok. But why is `track_running_stats` here?

**Review by VitalyFedyunin — APPROVED** [review]
